src/board_controller/biolistener/biolistener.cpp

```diff
@@ -0,0 +1,531 @@
+        }
+
+        keep_alive = false;
+        streaming_thread.join ();
+        return (int)BrainFlowExitCodes::STATUS_OK;
+    }
+    return (int)BrainFlowExitCodes::STREAM_THREAD_IS_NOT_RUNNING;
+}
+
+template <size_t BIOLISTENER_DATA_CHANNELS_COUNT>
+int BioListener<BIOLISTENER_DATA_CHANNELS_COUNT>::release_session ()
+{
+    if (initialized)
+    {
+        if (keep_alive)
+        {
+            stop_stream ();
+        }
+        initialized = false;
+        free_packages ();
+        if (control_socket)
+        {
+            control_socket->close ();
+            delete control_socket;
+            control_socket = NULL;
+        }
+        control_port = -1;
+        data_port = -1;
+    }
+    return (int)BrainFlowExitCodes::STATUS_OK;
+}
+
+template <size_t BIOLISTENER_DATA_CHANNELS_COUNT>
+bool BioListener<BIOLISTENER_DATA_CHANNELS_COUNT>::parse_tcp_buffer (const char *buffer,
+    size_t buffer_size, data_packet<BIOLISTENER_DATA_CHANNELS_COUNT> &parsed_packet)
+{
+    // Ensure the buffer size matches the packet size
+    if (buffer_size != packet_size)
+    {
+        safe_logger (spdlog::level::trace, "Buffer size mismatch!");
+
+        return false;
+    }
+
+    // Copy the raw buffer into the struct
+    std::memcpy (&parsed_packet, buffer, packet_size);
+
+    // Validate the header and footer
+    if (parsed_packet.header != BIOLISTENER_DATA_PACKET_HEADER)
+    {
+        safe_logger (spdlog::level::trace, "Invalid header! Expected: 0xA0");
+        return false;
+    }
+
+    if (parsed_packet.footer != BIOLISTENER_DATA_PACKET_FOOTER)
+    {
+        safe_logger (spdlog::level::trace, "Invalid footer! Expected: 0xC0");
+        return false;
+    }
+
+    return true;
+}
+
+
+// Convert raw ADC code (two's complement) to voltage
+// ref: reference voltage (X if milli volts output is needed, X * 1000 if micro volts output)
+//      1.2V for ADS131M08, 2.5V for AD7771
+// raw_code: raw ADC code
+// pga_gain: gain of the PGA
+// adc_resolution: resolution of the ADC (2^23 for ADS131M08, 2^24 for AD7771)
+template <size_t BIOLISTENER_DATA_CHANNELS_COUNT>
+double BioListener<BIOLISTENER_DATA_CHANNELS_COUNT>::data_to_volts (
+    double ref, uint32_t raw_code, double pga_gain, double adc_resolution)
+{
+    // Calculate the resolution in millivolts
+    double resolution = ref / (adc_resolution * pga_gain);
+
+    // Compute the voltage in millivolts based on the raw ADC code
+    if (raw_code <= 0x7FFFFF)
+    { // Positive range
+        return resolution * (double)raw_code;
+    }
+    else
+    { // Negative range (two's complement)
+        return (resolution * (double)raw_code) - (ref / pga_gain);
+    }
+}
+
+
+template <size_t BIOLISTENER_DATA_CHANNELS_COUNT>
+void BioListener<BIOLISTENER_DATA_CHANNELS_COUNT>::read_thread ()
+{
+    char message[sizeof (data_packet<BIOLISTENER_DATA_CHANNELS_COUNT>)];
+    int num_rows = board_descr["default"]["num_rows"];
+    double *package = new double[num_rows];
```

Comment:
I would rather create two packages, one per preset to keep it cleaner

Reply:
ok, so currently there is a builtin assumption that size of aux package is the same as default package